### PR TITLE
bpo-28879 : add date if missing in smtplib.send_message

### PR DESCRIPTION
**Pull request by elafontaine**
Hi I've tried to change the patch submitted on bpo-28879 to be on a github PR.

It's my first time contributing through PR, please let me know what I did wrong.

Regards,
Eric Lafontaine

https://bugs.python.org/issue28879

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by elafontaine** [comment]
oh wow... this is really a nice setup you've got running there!  I see that some check failed for some trailing whitespace in my files... 

Unfortunatly, I seems to be unable to look at the log anymore :S.

Anyway, how could I see it again?

**Comment by elafontaine** [comment]
Ok got to find how to see it again.  Now to understand what "trailing whitespace" is supposed to be for python and for... C (but I didn't touch the C part...)

**Comment by Mariatta** [comment]
The clue is at this line in travis log: (https://travis-ci.org/python/cpython/jobs/252242241#L494)

```
library/smtplib.rst:499: trailing whitespace
```

There is an extra space after the word `no` in line 499 of `smtplib.rst`.
The space should be removed.

File: Lib/test/test_smtplib.py [hunk 1]
```diff
@@ -549,6 +549,36 @@ def testSendMessageMultipleResentRaises(self):
...
+        m['X-Peer'] = socket.gethostbyname('localhost')
+        mexpect = '%s%s\n%s' % (MSG_BEGIN, m.as_string(), MSG_END)
+        self.assertEqual(self.output.getvalue(), mexpect)
+        debugout = smtpd.DEBUGSTREAM.getvalue()
+        current_moment = email.utils.formatdate()
+        Date = re.compile(''.join(("\\\\nDate: ",current_moment[:16])),re.MULTILINE)
+        self.assertRegex(debugout, Date)
+
+
+
```
Comment (mlouielu) [inline Lib/test/test_smtplib.py]: This only need two blankline to spacing class and class

File: Lib/smtplib.py [hunk 2]
```diff
@@ -928,6 +928,12 @@ def send_message(self, msg, from_addr=None, to_addrs=None,
             header_prefix = 'Resent-'
         else:
             raise ValueError("message has more than one 'Resent-' header block")
+        
+        # RFC 5322 section 3.6, 4th Paragraph
+        if msg.get('Date',None) is None:
+            msg['Date'] = email.utils.formatdate()
+        
+
```
Comment (mlouielu) [inline Lib/smtplib.py]: This only need one blankline.

**Review by mlouielu — CHANGES_REQUESTED** [review]
Hello, welcome to CPython community, thanks for contributing with CPython, and here is some coding style issue with PEP8 that can be fixed in your PR!

**Comment by elafontaine** [comment]
I've followed the recommandation (thanks @Mariatta @mlouielu ).  How can I re-trigger the CI now that the commit seems to be in the PR?

**Comment by mlouielu** [comment]
@lafe3402 CI will automatically re-trigger when you push your commit to GitHub. 

Also, I would recommend you to read this article: [How to write a git commit message](https://chris.beams.io/posts/git-commit/). This article is very useful that it tells how to write a good commit message. For example, your second commit message would be much better to changed to `Addressed somebody's comments` or `Fix coding style blankline`.

**Comment by elafontaine** [comment]
yeah... I though that I would have to make another PR request... so I've put what I've done in the subsection of the commit (it's in the expand the commit message if you expand it...).

Thanks a lot for the feedback :) you're really pro-active on the comment and I appreciate that.

Anything else I could do to ease the process?

Regards,
Eric Lafontaine

File: Lib/smtplib.py [hunk 3]
```diff
@@ -928,6 +928,11 @@ def send_message(self, msg, from_addr=None, to_addrs=None,
             header_prefix = 'Resent-'
         else:
             raise ValueError("message has more than one 'Resent-' header block")
+
+        # RFC 5322 section 3.6, 4th Paragraph
+        if msg.get('Date',None) is None:
```
Comment (mlouielu) [inline Lib/smtplib.py]: Here missing a whitespace after the `,`.

File: Lib/test/test_smtplib.py [hunk 4]
```diff
@@ -549,6 +549,34 @@ def testSendMessageMultipleResentRaises(self):
...
         smtp.close()
 
+    def testSendMessageAddDateIfMissing(self):
+        m = email.mime.text.MIMEText('A test message')
+        m['From'] = 'foo@bar.com'
+        m['To'] = 'John'
+        m['CC'] = 'Sally, Fred'
+        m['Bcc'] = 'John Root <root@localhost>, "Dinsdale" <warped@silly.walks.com>'
+        smtp = smtplib.SMTP(HOST, self.port, local_hostname='localhost', timeout=3)
+        smtp.send_message(m)
```
Comment (mlouielu) [inline Lib/test/test_smtplib.py:563]: Maybe we can mock up the `email.utils.formatdate()` to ensure we have the same result to tests.

File: Lib/test/test_smtplib.py [hunk 5]
```diff
@@ -549,6 +549,34 @@ def testSendMessageMultipleResentRaises(self):
...
+        self.output.flush()
+        # The Resent-Bcc headers are deleted before serialization.
+        del m['Bcc']
+        del m['Resent-Bcc']
+        # Add the X-Peer header that DebuggingServer adds
+        m['X-Peer'] = socket.gethostbyname('localhost')
+        mexpect = '%s%s\n%s' % (MSG_BEGIN, m.as_string(), MSG_END)
+        self.assertEqual(self.output.getvalue(), mexpect)
+        debugout = smtpd.DEBUGSTREAM.getvalue()
+        current_moment = email.utils.formatdate()
```
Comment (mlouielu) [inline Lib/test/test_smtplib.py]: see above to mock this up.

Comment (elafontaine) [reply Lib/test/test_smtplib.py]: Thanks for the review :) I'll work on it tonight, my wife won't be home so I'll have nothing better to do :).

I'm not sure what you mean though on the "mock this up" but I guess you mean to use @mock(email.utils.formatdate) ?

I'll have to read to remember how to do it correctly if it's what you meant.

**Comment by elafontaine** [comment]
I'm sorry about this last re-commit, I can't run the test right now so I'm using the CI :$.

**Comment by mlouielu** [comment]
@lafe3402 I can't pass the unittest on my laptop, too. You will need to check about the failed problem.

**Comment by elafontaine** [comment]
@mlouielu , thanks for your patience, I had trouble finding some free time.

**Comment by elafontaine** [comment]
Hi @mlouielu,

Would you have some feedback?  Just curious to see if there's something better I could've had done.

Regards,

**Comment by elafontaine** [comment]
@mlouielu @bitdancer,

Could you let me know what to do to accelerate the closure of this ticket?

**Comment by bitdancer** [comment]
Find me some free time somewhere?  I'm sorry, I just don't have the cycles to look at this, nor will I any time soon :( :(

**Comment by elafontaine** [comment]
@bitdancer, Anything I could do to help you?  

I've been re-reading my question and I just don't want it to be misinterpreted as a "I want my request to get in!" kind of request.  I'm genuinely trying to help you guys.  I love python and I would love to contribute more.  

Also, David, are you the only one who can approve?  If yes, should I learn email modules and try to alleviate some of the pain?  

Again, sorry for asking here, but I would like more meat on how to contribute or how to learn what I need to contribute XD.

Regards,
Eric Lafontaine

**Comment by lafe3402** [comment]
Should I make a file in news.d ?

**Comment by lafe3402** [comment]
closing to re-open on a rebased project

**Pull request closed without merging**